doc/contract-update.md

```diff
@@ -0,0 +1,56 @@
+# Upgrading and Migrating contracts
+
+We consider two approaches:  *Upgrading* existing contracts to a new version but retaining
+state data as-is, and; *migrating* users to a completely new set of contracts.
+
+We would only migrate in the event of a key loss, which prevents upgrading contracts.
+
+## Upgrade contracts approach
+
+Our contracts are already mostly divided into contracts that contain data and contracts that
+contain logic. This is fortuitous for the pattern we are going to use.
+
+`State.sol` contains most of the state, with some in `Shield.sol`.
+This division could be made complete by moving the `proposeBlock` function out of `State.sol`
+and moving the `Shield.sol` data into `State.sol` (there may be a few other minor adjustments
+TBC). `State.sol` will then only contain storage data and getters/setters for the same.
+It will also need modifying so that the addresses of the logic contracts, which it allows to
+set state, can be updated by a registered contract containing an `upgrade` function (see later).
+
+After that is done, we will never upgrade the `State.sol` contract but will upgrade the
+stateless logic contracts. The logic contracts call `State.sol` and thus we avoid the use of
+`delegateCall`. If, in the future, `State.sol` becomes insufficient for our needs, another
+data contract can be deployed alongside, which adds the new storage variables needed.
+
+We will create another contract, which proxies calls to the logic contracts. This avoids users
+having to repoint their applications to the new contracts. The `Proxy.sol` contract will also
+contain the `upgrade` function.
+
+![contract interaction](./contract-upgrade.png)
+
+### upgrade function
+
+This function will, atomically, carry out the following actions:
+
+1. Repoint `State.sol`'s registered logic contracts to the new logic contracts;
+1. Repoint `Proxy.sol` to the new contracts;
+1. Change the addresses of the contracts registered with `State.sol` to the new contract addresses;
+1. Store the address of the old contracts and the block number at which the swap-over occurred in `State.sol`; this
+will be used to help nightfall applications parse historic events and calldata. This will be in the form of
+an array, added to each time the contracts are upgraded.
+1. Call `selfdestruct` on the old logic contracts (just to return gas);
+1. Emit an event advertising the upgrade;
+
+We will ensure that a unique private key is needed to call `upgrade`.
+
+Note: The nightfall applications will need to be updated so that they can sync events and calldata on startup,
```

Comment:
This is the big thing. We use the `calldata` of the previous contracts. I think that `selfdestructing` the old contract should be enough, as no other clients would be able to use the old contract thus "faking" a valid version of NF. I think we should point this to the Security Audit guys, just to be sure that exposing the underlying proxied address and selfdestructing the old one after an update is safe

Reply:
We should, but I see no option here because one needs the calldata to sync back to the correct L2 state.  If for some reason that data is not available then the only option is a new deploy of the entire thing.

Reply:
Discussed with @signorecello the solution to this data availability could be Polygon Avail and then in the code go for this data in Polygon Avail instead of the calldata.

Reply:
Yes we talked about this, although I'm not sure Avail will be _Avail_able (pun intended) for our mainnet launch... Not sure if we want to have that dependency 🤷